addons/product/models/product_template.py

```diff
@@ -938,6 +939,85 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
+        One list of values by attribute line
+        :return: a generator of product template values combinations (in the form of lists)
+        """
+        if not product_template_attribute_values_list:
+            return
+
+        all_exclusions = self._get_own_attribute_exclusions()
+        # The following dict uses product template attribute values as keys
+        # 0 means the value is acceptable, greater than 0 means it's rejected, it cannot be negative
+        # Bear in mind that several values can reject the same value and the latter can only be included in the
+        #  considered combination if no value rejects it.
+        # This dictionary counts how many times each value is rejected.
+        # Each time a value is included in the considered combination, the values it rejects are incremented
+        # When a value is discarded from the considered combination, the values it rejects are decremented
+        current_exclusions = defaultdict(int)
+        for exclusion in self._get_parent_attribute_exclusions(parent_combination):
+            current_exclusions[exclusion] += 1
+        included_values_ids = set()
+
+        # The following list reflects product_template_attribute_values_list
+        # For each attribute, instead of a list of values, it contains the index of the selected value
+        # -1 means no value has been picked for the attribute in the current (partial) combination
+        values_idx = [-1] * len(product_template_attribute_values_list)
+        # determines which attribute line we're working on
+        attribute_idx = 0
+
+        def current_value_id():
+            return product_template_attribute_values_list[attribute_idx][values_idx[attribute_idx]].id
+
+        while True:
+            if values_idx[attribute_idx] >= 0:
+                for value_to_include_back in all_exclusions[current_value_id()]:
+                    current_exclusions[value_to_include_back] -= 1
+                included_values_ids.discard(current_value_id())
+
+            # first we try to get to next value of current attribute
+            if values_idx[attribute_idx] < len(product_template_attribute_values_list[attribute_idx]) - 1:
+                # get to next value of current attribute line
+                values_idx[attribute_idx] += 1
+                
+                for value_to_exclude in all_exclusions[current_value_id()]:
+                    current_exclusions[value_to_exclude] += 1
+                included_values_ids.add(current_value_id())
+            elif attribute_idx != 0:
+                # reset current attribute and get to previous attribute
+                values_idx[attribute_idx] = - 1
+                attribute_idx -= 1
+                continue
+            else:
+                # we're done if we must reset first attribute
+                break
+
+            # test if current value is excluded by included values or the opposite
```

Comment:
replace "the opposite" by explaining it, because any of "current value", "excluded by", and "included" could be made opposite individually, and we could also take opposite of some combination of them.

Reply:
ok.